File: app/src/main/java/com/adasa/foodplan/data/local/dao/IngredientDao.kt
package com.adasa.foodplan.data.local.dao

import androidx.room.*
import com.adasa.foodplan.data.local.entity.IngredientEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface IngredientDao {
    @Query("SELECT * FROM ingredients ORDER BY name ASC")
    fun getAllIngredients(): Flow<List<IngredientEntity>>

    @Query("SELECT * FROM ingredients WHERE id = :id")
    suspend fun getIngredientById(id: String): IngredientEntity?

    @Query("SELECT * FROM ingredients WHERE name LIKE '%' || :query || '%'")
    fun searchIngredients(query: String): Flow<List<IngredientEntity>>

    @Upsert
    suspend fun upsertIngredient(ingredient: IngredientEntity)

    @Upsert
    suspend fun upsertIngredients(ingredients: List<IngredientEntity>)

    @Delete
    suspend fun deleteIngredient(ingredient: IngredientEntity)
}

File: flutter-app/lib/data/local/dao/ingredient_dao.dart
import 'package:drift/drift.dart';
import 'package:foodplan/data/local/app_database.dart';

part 'ingredient_dao.g.dart';

@DriftAccessor(tables: [IngredientsTable])
class IngredientDao extends DatabaseAccessor<AppDatabase>
    with _$IngredientDaoMixin {
  IngredientDao(super.db);

  /// Watch all ingredients ordered by name.
  Stream<List<IngredientsTableData>> watchAllIngredients() =>
      (select(ingredientsTable)
            ..orderBy([(t) => OrderingTerm.asc(t.name)]))
          .watch();

  /// Watch ingredients whose name contains [query] (case-insensitive LIKE).
  Stream<List<IngredientsTableData>> searchIngredients(String query) =>
      (select(ingredientsTable)
            ..where((t) => t.name.like('%$query%'))
            ..orderBy([(t) => OrderingTerm.asc(t.name)]))
          .watch();

  /// One-shot fetch of a single ingredient by id.
  Future<IngredientsTableData?> getIngredientById(String id) =>
      (select(ingredientsTable)..where((t) => t.id.equals(id)))
          .getSingleOrNull();

  /// Upsert a single ingredient row.
  Future<void> upsertIngredient(IngredientsTableCompanion row) =>
      into(ingredientsTable).insertOnConflictUpdate(row);

  /// Upsert a batch of ingredient rows.
  Future<void> upsertIngredients(List<IngredientsTableCompanion> rows) async {
    await batch((b) {
      b.insertAllOnConflictUpdate(ingredientsTable, rows);
    });
  }

  /// Delete an ingredient by id.
  Future<void> deleteIngredientById(String id) =>
      (delete(ingredientsTable)..where((t) => t.id.equals(id))).go();
}
